text/0000-reformed-fn-pointers.md

```diff
@@ -0,0 +1,96 @@
+- Feature Name: reformed_fn_pointers
+- Start Date: 2015-03-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Repurpose Rust's current function pointer types to denote "incomplete function items", and introduce function reference types and new function pointer types, so that function references/pointers work more like value references/pointers.
+
+This is based on [RFC 883](https://github.com/rust-lang/rfcs/pull/883) and related discussions, where the following design is already largely agreed upon, but the author of RFC 883 doesn't have time to revise it. Therefore, this RFC is created to push for the changes.
+
+# Motivation
+
+Currently in Rust, there are two kinds of types that represent functions:
+
+1. Function item types: the types of function items, `fn(arg_list) -> ret_type {foo}`s.
+2. Function pointer types: the types of pointers to functions, `fn(arg_list) -> ret_type`s.
+
+Though called "function pointers", the function pointer types are considered *value* types (not reference/pointer types) by the language, which has the following problems:
+
+1. Inconsistencies in the language, especially when it comes to FFI codes that deal with nullable foreign function pointers. 
+2. There is currently no easy way to express the lifetime constraint on a function pointer to a non-`'static` function, which in particular makes it harder to implement type-safe JIT-ed functions/hot-loaded plugins in the future.
+
+Thus, this RFC proposes the following solution.
+
+# Detailed design
+
+Repurpose current function pointer types to denote "incomplete function items". Introduce function reference types of the form `&fn(arg_list) -> ret_type` and new function pointer types of the form `*const fn(arg_list) -> ret_type`.
+
+In the following section, `fn{f}`s, `fn`s, `&fn`s and `*const fn`s denote function item types, current function pointer types, function reference types and new function pointer types, respectively. Those types are considered "compatible" if their `(arg_list) -> ret_type` parts match.
+
+The following rules will apply:
+
+1. `fn{f}`s are still function item types, or "function handles/proxies/values", whose representations and semantics remain unchanged.
+2. `fn`s are no longer function pointer types, but types representing "incomplete function items", which are unsized statically and zero-sized dynamically.
+3. `&fn`s are function reference types, DST pointers with auxiliary data of type `()`, and work like other references.
+4. `*const fn`s are raw function pointer types that work as normally expected, and casting between compatible `&fn`s and `*const fn`s is valid.
+5. There are no `&mut fn`s, also no `*mut fn`s.
+6. `fn{f}`s still implement the closure traits (`Fn`/`FnMut`/`FnOnce`).
+7. `fn`s still implement the closure traits, for keeping `&fn`s coercible to closure trait objects.
+8. `&fn`s will implement the closure traits.
```

Comment:
no, I don't think `&fn` should implement the closure traits, just `fn`. In particular, we are likely to add blanket impls so that `&F` implements `Fn` for all `F`, but also this is not how any other callable things work (e.g. closures, trait objects, etc).